### PR TITLE
fix: false-positive DuplicatedTypeName for same-origin types

### LINKED ISSUE #4142: Duplicated type name error when running a single test
## Describe the Bug

In the project I'm working on we have a lots of types and we use strawberry.lazy very often. The application is functioning correctly, but sometimes there is a problem with running a single test (running all the tests works fine). An error is raised that some type is defined multiple times in the schema, but the error message indicates the same type twice. 

Simplified error message:

```
  error: Type `UserFilter` is defined multiple times in the schema              
                                                                                
       @ core_api/core/schema/users/users_types.py:133                          
                                                                                
    132 | @strawberry_django.filter_type(User, lookups=True)                    
  ❱ 133 | class UserFilter:                              
                ^^^^^^^^^^ first class defined here                             
    134 |     email: auto                                                                                                               
                                                                                
       @ core_api/core/schema/users/users_types.py:133                          
                                                                                
    132 | @strawberry_django.filter_type(User, lookups=True)                    
  ❱ 133 | class UserFilter:                              
                ^^^^^^^^^^ second class defined here                            
    134 |     email: auto                                                                                                                  
                                                                                
                                                                                
  To fix this error you should either rename the type or remove the duplicated  
  definition.                                                                   
                                                                                
  Read more about this error on                                                 
  https://errors.strawberry.rocks/duplicated-type-name     
```

## System Information

 - Operating system: MacOS 14.8
 - Python version: 3.12.8
 - Strawberry version (if applicable): 0.288.2

### PR DESCRIPTION
**Pull request by bellini666**
- Add `origin is origin` identity check in `is_same_type_definition` so that two different `StrawberryObjectDefinition` instances sharing the same origin class are recognized as the same type
- Prevents false-positive `DuplicatedTypeName` when third-party decorators (e.g. strawberry-django's `filter_type`) re-process a type, creating a new definition while keeping the same Python class as `origin`

Fix #4142

## Summary by Sourcery

Avoid treating multiple Strawberry object definitions with the same origin class as duplicate types and extend tests to cover this behavior.

Bug Fixes:
- Prevent false-positive DuplicatedTypeName errors when separate StrawberryObjectDefinition instances share the same origin class.

Tests:
- Add regression tests ensuring same-origin type definitions are treated as the same type while genuinely different types with the same name still raise duplication errors.

### CODE REVIEW

**Comment by botberry** [comment]
Thanks for adding the `RELEASE.md` file!



Here's a preview of the changelog:

---

Fix false-positive `DuplicatedTypeName` error when two different `StrawberryObjectDefinition` instances share the same `origin` class. This can happen when third-party decorators (e.g. strawberry-django's `filter_type`) re-process a type, creating a new definition while keeping the same Python class as origin.



Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to @_bellini666 for the PR 👏

Get it here 👉 https://strawberry.rocks/release/(next)
```

File: tests/schema/test_duplicated_types.py [hunk 1]
```diff
@@ -261,3 +262,57 @@ class Query:
...
+
+    @strawberry.type(name="Shared")
+    class B:
+        b: int
+
+    @strawberry.type
+    class Query:
+        field: int
+
+    strawberry.Schema(query=Query, types=[A, B])
```
Comment (patrick91) [inline tests/schema/test_duplicated_types.py]: Don't we have a test like this already?😊

File: tests/schema/test_duplicated_types.py [hunk 2]
```diff
@@ -261,3 +262,57 @@ class Query:
...
+    class Query:
+        foo: Foo
+
+    schema = strawberry.Schema(query=Query)
+    converter = schema.schema_converter
+
+    # "Foo" is already cached from schema construction via Query.foo.
+    # Calling from_object with a *different* definition instance (same origin)
+    # must not raise DuplicatedTypeName.
+    converter.from_object(duplicate_def)
```
Comment (patrick91) [inline tests/schema/test_duplicated_types.py]: I'd prefer to have a real test that exercises this edge case, might be more complicated, but I want to avoid these kinds of tests 😊

**Comment by botberry** [comment]
# Pre-release
:wave:

Pre-release **0.291.2.dev1771437961** [fc0004eeb079bb099f6c22f88fcbdf2e053a9e8a] has been released on PyPi! :rocket:
You can try it by doing:

```shell
poetry add strawberry-graphql==0.291.2.dev1771437961
```

**Comment by botberry** [comment]
# Pre-release
:wave:

Pre-release **0.291.2.dev1770456508** [a98d9df2bea5d6d2264e579242b161c4a69d64cc] has been released on PyPi! :rocket:
You can try it by doing:

```shell
poetry add strawberry-graphql==0.291.2.dev1770456508
```
